### PR TITLE
Add API docs

### PR DESCRIPTION
**Pull request by kcpevey**
## Changes introduced in this PR:

- add api doc md file
- add script to generate md file
- add gh workflow to autogenerate docs

## Reference Issues or PRs

Partially resolves https://github.com/nebari-dev/nebari-docs/issues/157

## What does this implement/fix?

- [x] Documentation Update

## Testing

- [x] Did you test the pull request locally?
- [ ] Did you add new tests?

## Any other comments?

There is a sister PR on the nebari-docs repo to add it this actual file to the documentation. I still need to add a separate GH action to open the PR on the nebari-docs repo

This PR supercedes [this one](https://github.com/nebari-dev/nebari/pull/1511) which was failing due to credentials access via a fork.

### CODE REVIEW

File: .github/workflows/update_api_docs.yml
Comment (iameskild) [inline .github/workflows/update_api_docs.yml:6]: I wonder if we should trigger this workflow when commits are pushed to `develop`, what are your thoughts?

File: docs/api_docs/api_doc.md
Comment (iameskild) [inline docs/api_docs/api_doc.md:1]: Just to confirm, this is the doc that will be merged into the nebari-docs repo right?

Comment (iameskild) [reply docs/api_docs/api_doc.md:1]: And what are the differences between the [PR opened](https://github.com/nebari-dev/nebari/pull/1635/files) on this repo by the Nebari-Sensei bot and this[ PR opened](https://github.com/nebari-dev/nebari-docs/pull/190/files) on the docs repo?

**Review by iameskild — COMMENTED** [review]
Hi @kcpevey I left a few questions and comments but otherwise this is looking good. 

I think having API docs will not only be useful for newcomers but for us as maintainers as well. Reviewing the API docs makes it more obvious where we need to make improvements :)

File: .github/workflows/update_api_docs.yml
Comment (kcpevey) [reply .github/workflows/update_api_docs.yml:6]: I'm not sure what the git pattern is these days. Whatever makes sense is fine with me.

File: docs/api_docs/api_doc.md
Comment (kcpevey) [reply docs/api_docs/api_doc.md:1]: yes, this doc is what will go into the actual docs. 

The PR on the docs repo is outdated at this point. After discussion with the team, I need to add an additional GH action to trigger the creating of a PR over on the docs repo.

**Comment by iameskild** [comment]
@kcpevey do you think this will be wrapped up this week or should we push it to the next release?

**Comment by kcpevey** [comment]
@iameskild this is almost ready. We discussed using the sync workflow that @pavithraes already has to keep the files in sync rather than writing something new. 

I need to know what branch(es) to trigger this from https://github.com/nebari-dev/nebari/pull/1634#discussion_r1104920361

**Comment by iameskild** [comment]
Sounds good, thanks @kcpevey!

I see this workflow being triggered when changes are _merged into the following branches_ (either one or possibly both):
 - `develop` if we want to keep the API docs up-to-date with all of the latest changes, and/or
 - `main` if we want to keep the API docs up-to-date with the latest stable release.

**Comment by kcpevey** [comment]
Last I looked at this, it was ready for review.

File: .github/workflows/generate_cli_doc.yml
Comment (iameskild) [inline .github/workflows/generate_cli_doc.yml:7]: This will only update the API docs after each release; that seems appropriate. I'm curious if we want to also consider doing so after each merge to `develop`. What are your thoughts?

**Review by iameskild — APPROVED** [review]
I tested this as much as far as I could on my local machine and I was able to render the API docs `cli.html` and they look good to me. I left one minor comment but otherwise, I think this is ready to go! Thanks @kcpevey

File: .github/workflows/generate_cli_doc.yml
Comment (kcpevey) [reply .github/workflows/generate_cli_doc.yml:7]: I don't think generating docs with every merge to develop would be correct here. We only have one version of the docs for the release branch so I wouldnt want to confuse people by adding features which arent yet released into the docs.

**Comment by kcpevey** [comment]
I will try to bring this up to date with the CI fixes on develop this week.

**Comment by iameskild** [comment]
Hi @kcpevey, this should be ready to merge right? CI is failing at the moment because the DigitalOcean token has expired (I lack the access at the moment to update it) but once the kubernetes test passes, I'm good with merging this :)

**Comment by iameskild** [comment]
@kcpevey is this PR ready to merge? If you're busy, I can wrap this up.

**Comment by kcpevey** [comment]
This should be ready to go.

**Comment by iameskild** [comment]
Hey @kcpevey, do you have time this week to resolve the merge conflicts? Once that's done, I believe this can be merged :)

**Comment by iameskild** [comment]
Hey @kcpevey, this looks like it's really close, do you have a few minutes to help resolve this conflicts? Once that's done, we can merge this PR.

**Comment by kcpevey** [comment]
I struggled a good bit trying to get Typer to render like I wanted. The sphinx extension `sphinx-click` does a good job, but there are still two issues that have no resolution at this point. 

1. The rich markdown coloring of the help docs is not understood by sphinx-click. This is a known issue marked low priority in sphinx-click (https://github.com/ewels/rich-click/issues/48).
2. The name of the program appears to be main. I tried renaming the files, setting `prog_name` (https://github.com/tiangolo/typer/pull/120), renaming the imports, and a bunch of other things but didn't find a solution. 

I'm not sure how people feel about these "warts". I could programmatically do some find/replace on the output html, or we can accept as is. 

image

**Pull request merged**
